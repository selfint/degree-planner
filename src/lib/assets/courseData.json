{"97030013":{"code":"97030013","about":"","points":0,"name":"מתמטיקה מכינה בינלאומי","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter"]},"97030014":{"code":"97030014","about":"","points":0,"name":"פיזיקה מכינה בינלאומי","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter"]},"97030015":{"code":"97030015","about":"","points":0,"name":"עברית מכינה בינלאומי","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter"]},"03940902":{"code":"03940902","median":98.5,"about":"הציון ייקבע עפ\"י מעקב והערכה במשך הסמסטר ו.או מבחני מיומנות. הרישום לנבחרת מתבצ ע ע\"י המאמן לאחר בחינת רמתו הספורטיבית של המועמד במהלך שלושת השיעורים הראשונים של הסמסטר. סטודנט שיתקבל לאחת הנבחרות או קורסי המתקדמים חייב לבטל מיידית את ריש ומו למקצוע ספורט אחר, אחרתלא ייקלט רישומו. סטודנט אינו רשאי להירשם ליותר מאחד מ מקצועות החינוך הגופני והנבחרות באותו סמסטר. נוכחות חובה בכל השיעורים. להלן פירו ט הנבחרות: אתלטיקה קלה, כדור עף - אולמות וחופים, כדור סל, כדור רגל, טניס, טניס שולחן, ריצות ארוכות, כדור יד, ג'ודו, סייף, שייט, שחמט, קראטה, סלסה - להקה, שחיי ה, כדור מים, להקת מחול, סקווש, ניווט, אופני הרים, באולינג, גלשני מפרש, ו, קארטינג, דיבייט, קיאקים,  סנונית, להקת מעודדות.","points":1.5,"name":"נבחרות ספורט","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940820":{"code":"03940820","median":97,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - תנועה ומחול","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring","Summer"]},"03940808":{"code":"03940808","median":95,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני-מיועד לסטודנטים חדשים","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940807":{"code":"03940807","median":96.5,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"כושר גופני מעורב","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring","Summer"]},"03940806":{"code":"03940806","median":96,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - הגנה עצמית","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940805":{"code":"03940805","median":98,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - אתלטיקה קלה","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940804":{"code":"03940804","median":95.25,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - משחקי מחבט","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940802":{"code":"03940802","median":95.2,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - שחיה","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940801":{"code":"03940801","median":95.75,"about":"סטודנטים חייבים לצבור במהלך לימודיהם, שתי נקודות מתוך רשימת קורסי חינוך גופני,  בהם מלמדים תאוריה ומתרגלים מיומנויות.\nנוכחות חובה בשיעורים. הציון יקבע על סמך מעקב והערכה במשך הסמסטר ו.או מבחני מיומנויות. \nאחריות הרישום והביטול לקורסים  חלה על הסטודנט.ית בלבד.\nלא ניתן להירשם \\ להשתתף ליותר מקורס או נבחרת באותו סמסטר.\n","points":1,"name":"חינוך גופני - התעמלות כללית בנות","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring"]},"03940582":{"code":"03940582","median":100,"about":"לימוד מיומנות בנגינה תזמורתית והבאת יצירות סימפוניות לביצוע. הכנת רפרטואר מגוון כולל ליווי סולנים ומקהלה. מתקבלים רק נגנים בכלי תזמורת, ועל סמך בחינה מעשית. חובה להשתתף החל מהשיעור הראשון.","points":2,"name":"תזמורת","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring","Summer"]},"03360024":{"code":"03360024","median":93,"about":"הסטודנטים יבחרו, יאפיינו ויבססו צורך קליני ויחלו בסיעור מוחות לגבי פתרונות טכנולוגיים אפשריים. בהמשך, הסטודנטים יעריכו את הפתרונות הטכנולוגיים שהוצעו ויבחרו בפתרון מוביל  אותו יפתחו באמצעות בניית אבותי- טיפוס, הפחתת הסיכון הטכני, אסטרטגיות לטיפול בדרישות ספציפיותלבריאות (רגולציה, שיפוי), ואסטרטגיה עסקית (קניין רוחני, מימון, מסחור). \r\n תוצאות למידה: הקורס מספק מבוא אינטנסיבי לתהליך של חדשנות טכנולוגית בתחום הרפואי. כחלק מצוות רב-תחומי פרוייקט, ילמד התלמיד:\r\n כיצד לזהות צורך קליני משמעותי ללא פתרון קיים.\r\n לתכנן פתרון טכנולוגי ולייצר אב-טיפוס כדי לענות על הצורך הרפואי שזוהה.\r\n להכין תכנית להבאת הפתרון לשוק (כולל התיחסות גיוס מימון, קניין רוחני, רגולציה, שיפ וי ומודלים עסקיים).\r\n","points":3,"name":"חדשנות רפואית במודל ביודזיין 1","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter"]},"03240567":{"code":"03240567","median":94.6,"about":"סדנה זו תתמקד בהתפתחות יכולת ההסתכלות והיישום שלה לצילום, שפת הצילום והטכניקות שלה כגון אור, נקודת המבט, תזמון וקומפוזיציה, יחד עם תיקון ומניפולציה של התמונה. הסדנה תשלב תרגול מונחה בכיתה ובחוץ ותציג מגוון של נושאים הכוללים דיוקן, דיוקן עצמי, טבע דומם, צילומי טבעוצילומי סביבה עירונית.","points":1.5,"name":"סדנת צילום: שפה וכלים","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring","Summer"]},"03240483":{"code":"03240483","median":95.4,"about":"מטרת הקורס: פיתוח יכולת ראשונית לציור המעשי, לימוד תיאורטי של מערכת המושגים והעקרונות באמנות הציור, הקניית יכולת בסיסית לציור אובייקטים שונים: צורות בסיסיות, דומם, אובייקטים מורכבים (נוף), הצגת אור וצל, פרספקטיבה אווירית, קומפוזיציה, וכו'. לימוד הטכניקה של אקוורל(צבע מים).","points":2,"name":"ציור למתחילים","connections":{"dependencies":[],"adjacent":[],"exclusive":[]},"seasons":["Winter","Spring","Summer"]}}
