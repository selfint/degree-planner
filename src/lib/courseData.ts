const cacheRoute = '/_cache/courseData/';
const courseData: CourseData = new Map();

export async function getAllCourses(): Promise<Course[]> {
	return await Promise.all(Array.from(courseData.values()));
}

export function getAllCoursesSync(): Promise<Course[]> {
	return Promise.all(Array.from(courseData.values()));
}

export function courseCodeIsValid(code: string): boolean {
	return /^\d{8}$/.test(code);
}

export function buildGetCourseData() {
	const abort = new AbortController();
	const abortSignal = abort.signal;

	return {
		getCourseData: (code: string) => getCourseData(code, { abortSignal }),
		abort: () => abort.abort()
	};
}

export async function getCourseDataBatch(
	codes: string[],
	opts?: { abortSignal?: AbortSignal }
): Promise<Course[]> {
	const newCodes = codes.filter((c) => !courseData.has(c));
	const oldCodes = codes.filter((c) => courseData.has(c));

	const serverRes = await fetch(`/api/courseInfo`, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json'
		},
		body: JSON.stringify(newCodes),
		signal: opts?.abortSignal
	});

	const cacheRes = Promise.all(
		// these codes are all in the courseData map
		oldCodes.map((c) => courseData.get(c) as Promise<Course>)
	);

	let courses: Course[];
	if (serverRes.ok) {
		courses = await serverRes.json();
	} else {
		return await cacheRes;
	}

	for (const course of courses) {
		courseData.set(course.code, Promise.resolve(course));
	}

	return courses.concat(await cacheRes);
}

/**
 * Get course data for a course code
 * @param code Course code
 * @param opts options - abortSignal: AbortSignal to cancel the request
 * @returns Course object, will return only the course code if the course is not found
 */
export function getCourseData(
	code: string,
	opts?: { abortSignal?: AbortSignal }
): Promise<Course> {
	if (!courseCodeIsValid(code)) {
		return Promise.resolve({ code });
	}

	const data = courseData.get(code);
	if (data === undefined) {
		let done = false;
		async function getData(): Promise<Course> {
			// Fetch first from cache, then from server if cache fails
			try {
				const cacheRes = await fetch(cacheRoute + code + '.json', {
					signal: opts?.abortSignal
				});
				if (cacheRes.ok) {
					done = true;
					return cacheRes.json();
				}
			} catch (e) {
				//
			}

			try {
				const serverRes = await fetch(`/api/courseInfo/${code}`, {
					signal: opts?.abortSignal
				});

				if (serverRes.ok) {
					done = true;
					return serverRes.json();
				}
			} catch (e) {
				//
			}

			return {
				code
			};
		}

		const future = getData()
			// start fetching dependencies
			.then((course) => {
				course.connections?.dependencies.forEach((dep) => {
					dep.forEach((code) => {
						getCourseData(code);
					});
				});

				return course;
			});
		courseData.set(code, future);

		if (opts?.abortSignal) {
			opts.abortSignal.addEventListener('abort', () => {
				if (!done) {
					courseData.delete(code);
				}
			});
		}

		return future;
	} else {
		return data;
	}
}

export function cacheDegreeCourses(degreeData: DegreeData): void {
	const courses = degreeData.recommended.flat();

	courses.push(
		...new Set(
			Array.from(degreeData.requirements.requirements.values()).flatMap(
				getRequirementCourses
			)
		)
	);

	courses.forEach((code) => {
		getCourseData(code);
	});
}

function getRequirementCourses(requirement: Requirement): string[] {
	if (requirement.courses !== undefined) {
		return requirement.courses;
	}

	if (requirement.choice !== undefined) {
		let courses: string[] = [];
		for (const [_, option] of requirement.choice.options) {
			courses.push(...getRequirementCourses(option));
		}

		return courses;
	}

	return [];
}
