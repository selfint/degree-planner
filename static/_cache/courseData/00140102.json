{"code":"00140102","points":4.5,"name":"מבוא למכניקה הנדסית","connections":{"dependencies":[["00140010"],["02050691"],["02750103"]],"adjacent":[],"exclusive":[]},"about":"מבוא! וקטור כח! וקטור מומנט. מערכות שקולות במישור ובמרחב! מערכת שקולה פשוטה ביותר. מושג הגוף החופשי! תנאי שווי משקל! מושגי היציבות והסיום הסטטי. מסבכים מישוריים ומרחביים! קורות ומסגרות מישוריות ומרחביות. עומסים מפורשים. מהלכי כוחות ומומנטים פנימיים במבנים. קשרים דיפרנציאליים לכוחות פנימיים במבנים. שיטת הסופרפוזיציה לחישוב מהלכי כוחות פנימיים. מבנה כבלים וקשתות. תכונות גופים ושטחים )מרכזית! מומנט ראשון ושני של השטח(! מרכז כובד ומרכז מסה.","tests":[{"year":2025,"monthIndex":2,"day":4},{"year":2025,"monthIndex":2,"day":28}],"seasons":["Winter"]}