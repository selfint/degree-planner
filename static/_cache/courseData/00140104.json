{"code":"00140104","median":65.375,"points":4,"name":"תורת החוזק 1","connections":{"dependencies":[["00140102"]],"adjacent":[],"exclusive":[]},"about":"מבוא. מאמץ ועיבור במשיכה ולחיצה. פיתול חתכים שונים, כפיפת קורות אלסטיות והומוגניות. גזירה בשעת כפיפה, מרכז הגזירה. טרנספורמצית המאמץ והעיבור, שקיעות בעזרת המשוואה הדיפרצניאלית של הקו האלסטי, פתרון מבנים בלתי מסוימים. מושגים ראשונים של קריסה אלסטית.","tests":[{"year":2025,"monthIndex":1,"day":12},{"year":2025,"monthIndex":2,"day":18}],"seasons":["Winter","Spring","Summer"]}