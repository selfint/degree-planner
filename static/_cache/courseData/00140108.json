{"code":"00140108","median":73.7,"points":3,"name":"סטטיקת מבנים","connections":{"dependencies":[["00140104"]],"adjacent":[],"exclusive":[]},"about":"מהלכי כוחות פנימיים וקווי השפעה במבנים מסוימים סטטית. משפט בטי וחישוב הזזות אלסטיות. שיטת הגמישות. שיטת הקשיחות. קווי השפעה במבנים בלתי-מסוימים סטטית.","tests":[{"year":2025,"monthIndex":1,"day":20},{"year":2025,"monthIndex":2,"day":20}],"seasons":["Winter","Spring","Summer"]}