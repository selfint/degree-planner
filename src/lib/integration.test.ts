import fs from 'fs/promises';
import path from 'path';
import { describe, it } from 'vitest';

import { getProgress, requirementCompleted } from './progress';
import { loadCatalog, parseCatalog } from './requirements';
import { getCourseData } from './courseData';

async function localFetch(
	input: string | URL | globalThis.Request,
	init?: RequestInit
): Promise<Response> {
	// assert input is string
	if (typeof input !== 'string') {
		throw new Error('Only string input is supported');
	}

	const base = path.join(process.cwd(), 'static');
	const filepath = path.join(base, input);

	// read the file
	const text = await fs.readFile(filepath).then((buf) => buf.toString());

	return new Response(text);
}

describe('Integration', () => {
	it('should work with CS degree', async (ctx) => {
		const degree: Degree = ['2023_2024', 'computer_science', '3_year'];

		const data = await loadCatalog(degree, localFetch);

		const progress = getProgress(
			data.recommended.map((s) => s.map(getCourseData)),
			data.requirement
		);

		ctx.expect(progress).toMatchSnapshot();
	});
});

describe('CS 4 year degree', () => {
	it('should only count specialization points from completed specializations', async (ctx) => {
		const degree: Degree = ['2023_2024', 'computer_science', '4_year'];

		function buildSemesters(semesters: string[][]): Course[][] {
			return semesters.map((semester) => semester.map(getCourseData));
		}

		const data = await loadCatalog(degree, localFetch);

		const specialization = data.requirement.nested
			?.find((r) => r.name === 'list_a')
			?.nested?.find((r) => r.name === 'specialization');

		ctx.expect(specialization?.strict).toEqual('points');

		const chosenSpecializations = parseCatalog(`
			1 סיבוכיות של חישובים נק'
			236306 גרפים מקריים 2.0
			236309 מבוא לתורת הצפינה 3.0
			236313 תורת הסיבוכיות 3.0


			2 תורת האלגוריתמים
			236521 אלגוריתמי ק י רוב 2.0
			236377 אלגוריתמים מבו זרים בגרפים 3.0
			236359 אלגוריתמים 2 3.0


			3 לוגיקה ויישומיה
			236025 אוטומטים, לוגיקה ומשחקים 2.0
			236026 ידע ומשחקים במערכות מבוזרות 2.0
			236304 לוגיקה למדעי המחשב 2 3.0
		`);

		const other = parseCatalog(`
			5 פיתוח מערכות תוכנה
			236268 ארכיטקטורת מעבדים בגישה בונה 3.0
			236271 פיתוח מבוסס אנדרואיד 2.0
		`);

		ctx.expect(chosenSpecializations).toHaveLength(9);
		ctx
			.expect([...new Set([...chosenSpecializations, ...other])].length)
			.toEqual(11);

		const expectedPoints = chosenSpecializations
			.map(getCourseData)
			.reduce((acc, course) => acc + (course.points ?? 0), 0);

		ctx.expect(expectedPoints, 'Points to match').toEqual(23);

		const progress = getProgress(
			buildSemesters([chosenSpecializations, other]),
			specialization!
		);

		ctx.expect(progress.points).toMatchInlineSnapshot(`
			{
			  "done": 23,
			  "required": 26,
			}
		`);
		ctx.expect(progress.count).toMatchInlineSnapshot(`
			{
			  "done": 9,
			  "required": 0,
			}
		`);
		ctx.expect(progress.amount).toMatchInlineSnapshot(`
			{
			  "done": 3,
			  "required": 3,
			}
		`);
		ctx.expect(progress.nested.done.map(({ name }) => name))
			.toMatchInlineSnapshot(`
			[
			  "algorithm_theory",
			  "applied_logic",
			  "computation_complexity",
			]
		`);
		ctx.expect(requirementCompleted(progress)).toEqual(false);
	});
});
