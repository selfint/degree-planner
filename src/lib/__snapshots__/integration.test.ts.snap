// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration > should work with CS degree 1`] = `
{
  "points": [
    73.5,
    118.5,
  ],
  "requirements": Map {
    "core" => [
      {
        "courses": [
          "01040031",
          "01040166",
          "02340114",
          "02340129",
          "03240033",
          "01040032",
          "01140071",
          "02340124",
          "02340125",
          "02340141",
          "00940412",
          "01040134",
          "02340218",
          "00440252",
          "02340252",
          "02340292",
          "02340118",
          "02340123",
          "02340247",
          "01040135",
          "01040033",
          "01040174",
          "01040122",
          "01040142",
          "01040285",
          "01040295",
          "01140075",
          "01140052",
          "01140054",
          "01140073",
          "01140101",
          "01140246",
          "01240120",
          "01250001",
          "01250801",
          "01240510",
          "01340058",
          "01340020",
          "02360343",
          "02360360",
        ],
        "points": 84.5,
      },
      {
        "courses": [
          "01040031",
          "01040166",
          "02340114",
          "02340129",
          "03240033",
          "01040032",
          "01140071",
          "02340124",
          "02340125",
          "02340141",
          "00940412",
          "01040134",
          "02340218",
          "00440252",
          "02340292",
          "02340118",
          "02340123",
          "02340247",
          "02360343",
          "02360360",
        ],
        "points": 73.5,
      },
    ],
    "english" => [
      {
        "count": 2,
        "courses": [
          "00860762",
          "00840515",
          "00860320",
          "00860478",
          "00850915",
          "00860800",
          "00860822",
          "02740251",
          "01980000",
          "02360009",
          "00160102",
          "00160208",
          "00160328",
          "00170012",
          "00140603",
          "00140143",
          "00160503",
          "03150242",
          "03150014",
          "03240262",
          "03240282",
          "03240305",
          "03240527",
          "03250016",
          "03260001",
          "03260002",
          "03260006",
          "03260009",
          "01040122",
          "01040181",
          "01060350",
          "01060395",
          "01060429",
          "01060803",
          "01060935",
          "01060937",
          "01960012",
          "01960014",
          "01970007",
          "01970014",
          "00460200",
          "00460240",
          "00460746",
          "01340039",
          "01340128",
          "01360037",
          "01140229",
          "01140252",
          "01140250",
          "01140251",
          "00560149",
          "00560391",
          "00560404",
          "00560406",
          "00560407",
          "00560403",
          "00560405",
          "00560146",
          "03360014",
          "03360032",
          "03360209",
          "03360539",
          "03360541",
          "03360549",
          "03380001",
          "02360025",
          "02360306",
          "02360613",
          "02360640",
          "02360664",
          "02360824",
          "00190627",
          "02050586",
          "02050028",
          "02060926",
          "02070048",
          "02060006",
          "02060840",
          "00340034",
          "00360001",
          "00360004",
          "00360008",
          "00360010",
          "00360013",
          "00360014",
          "00360026",
          "00360027",
          "00360032",
          "00360035",
          "00360044",
          "00360063",
          "00360065",
          "00360079",
          "00360080",
          "00360711",
          "00360097",
          "00360103",
          "00660605",
          "00660532",
          "00660533",
          "01260200",
          "01260603",
          "01260700",
          "01270002",
          "01270446",
          "01270727",
          "00960226",
          "00960412",
          "00960606",
        ],
      },
      {
        "count": 0,
        "courses": [],
      },
    ],
    "list_a" => [
      {
        "courses": [
          "02340268",
          "02340302",
          "02340303",
          "02340304",
          "02340313",
          "02340326",
          "02340329",
          "02340493",
          "02340901",
          "02360002",
          "02360004",
          "02360025",
          "02360026",
          "02360201",
          "02360203",
          "02360204",
          "02360205",
          "02360206",
          "02360216",
          "02360268",
          "02360270",
          "02360271",
          "02360272",
          "02360278",
          "02360299",
          "02360303",
          "02360304",
          "02360306",
          "02360309",
          "02360310",
          "02360313",
          "02360315",
          "02360318",
          "02360319",
          "02360321",
          "02360322",
          "02360323",
          "02360324",
          "02360328",
          "02360329",
          "02360330",
          "02360332",
          "02360333",
          "02360334",
          "02360336",
          "02360340",
          "02360341",
          "02360342",
          "02360345",
          "02360346",
          "02360347",
          "02360348",
          "02360349",
          "02360350",
          "02360351",
          "02360356",
          "02360357",
          "02360358",
          "02360359",
          "02360360",
          "02360361",
          "02360363",
          "02360366",
          "02360369",
          "02360370",
          "02360371",
          "02360372",
          "02360373",
          "02360374",
          "02360376",
          "02360377",
          "02360378",
          "02360379",
          "02360381",
          "02360388",
          "02360422",
          "02360490",
          "02360491",
          "02360496",
          "02360499",
          "02360500",
          "02360501",
          "02360502",
          "02360503",
          "02360504",
          "02360506",
          "02360508",
          "02360509",
          "02360510",
          "02360512",
          "02360513",
          "02360515",
          "02360518",
          "02360520",
          "02360521",
          "02360522",
          "02360523",
          "02360524",
          "02360525",
          "02360526",
          "02360612",
          "02360613",
          "02360620",
          "02360621",
          "02360622",
          "02360623",
          "02360624",
          "02360625",
          "02360627",
          "02360628",
          "02360629",
          "02360630",
          "02360631",
          "02360632",
          "02360633",
          "02360634",
          "02360635",
          "02360637",
          "02360638",
          "02360640",
          "02360641",
          "02360643",
          "02360644",
          "02360645",
          "02360646",
          "02360647",
          "02360648",
          "02360649",
          "02360650",
          "02360651",
          "02360652",
          "02360653",
          "02360654",
          "02360655",
          "02360657",
          "02360658",
          "02360660",
          "02360661",
          "02360662",
          "02360663",
          "02360664",
          "02360667",
          "02360668",
          "02360669",
          "02360670",
          "02360698",
          "02360700",
          "02360703",
          "02360712",
          "02360715",
          "02360716",
          "02360719",
          "02360729",
          "02360754",
          "02360755",
          "02360756",
          "02360757",
          "02360760",
          "02360777",
          "02360779",
          "02360780",
          "02360781",
          "02360800",
          "02360811",
          "02360812",
          "02360813",
          "02360814",
          "02360815",
          "02360816",
          "02360817",
          "02360818",
          "02360819",
          "02360820",
          "02360821",
          "02360822",
          "02360823",
          "02360824",
          "02360825",
          "02360826",
          "02360827",
          "02360828",
          "02360829",
          "02360830",
          "02360831",
          "02360832",
          "02360833",
          "02360834",
          "02360835",
          "02360836",
          "02360837",
          "02360838",
          "02360860",
          "02360861",
          "02360862",
          "02360873",
          "02360874",
          "02360875",
          "02360901",
          "02360927",
          "02360990",
          "02360991",
          "02380100",
          "02380125",
          "02380739",
          "02380790",
          "02380900",
          "02380901",
          "02380902",
        ],
        "overflow": "list_b",
        "points": 18,
      },
      {
        "courses": [
          "02360360",
        ],
        "points": 3,
      },
    ],
    "list_b" => [
      {
        "courses": [
          "00360044",
          "00440105",
          "00440127",
          "00440131",
          "00440137",
          "00440157",
          "00440167",
          "00440169",
          "00440202",
          "00460201",
          "00460206",
          "00460332",
          "00460880",
          "00480878",
          "00480921",
          "00860761",
          "00940222",
          "00940313",
          "00940314",
          "00940333",
          "00940334",
          "00940423",
          "00940591",
          "00960200",
          "00960211",
          "00960224",
          "00960250",
          "00960262",
          "00960326",
          "00960411",
          "00970317",
          "01040122",
          "01040135",
          "01040142",
          "01040157",
          "01040165",
          "01040174",
          "01040158",
          "01040177",
          "01040192",
          "01040221",
          "01040223",
          "01040279",
          "01040293",
          "01060378",
          "01040294",
          "01060383",
          "01140101",
          "01140246",
          "01150203",
          "01150204",
          "01140036",
          "01160217",
          "01160354",
          "01240120",
          "01240400",
          "01240503",
          "01240801",
          "01250801",
          "01340019",
          "01340020",
          "01340058",
          "01340082",
          "01340113",
          "01340128",
          "01340119",
          "01340142",
          "02140909",
          "01040033",
          "01040285",
          "01040295",
        ],
        "points": 6.5,
      },
      {
        "courses": [],
        "points": 0,
      },
    ],
    "malag" => [
      {
        "courses": [
          "02140120",
          "03240258",
          "03240262",
          "03240265",
          "03240269",
          "03240273",
          "03240282",
          "03240284",
          "03240286",
          "03240292",
          "03240298",
          "03240305",
          "03240306",
          "03240307",
          "03240315",
          "03240436",
          "03240441",
          "03240442",
          "03240445",
          "03240474",
          "03240518",
          "03240520",
          "03240527",
          "03240528",
          "03240540",
          "03240541",
          "03240962",
          "03240987",
          "03240992",
          "03250001",
          "03250002",
          "03250006",
          "03250010",
          "03250011",
          "03250012",
          "03250015",
          "03250016",
          "03250017",
          "03250019",
          "03250020",
          "03260001",
          "03260002",
          "03260006",
          "03260009",
        ],
        "points": 6,
      },
      {
        "courses": [],
        "points": 0,
      },
    ],
    "math" => [
      {
        "count": 1,
        "courses": [
          "01040135",
          "01040033",
          "01040174",
          "01040122",
          "01040142",
          "01040285",
          "01040295",
        ],
      },
      {
        "count": 0,
        "courses": [],
      },
    ],
    "other" => [
      {
        "courses": [],
        "points": 2,
      },
      {
        "courses": [],
        "points": 0,
      },
    ],
    "project" => [
      {
        "count": 1,
        "courses": [
          "02340302",
          "02340303",
          "02340304",
          "02340313",
          "02340326",
          "02340329",
          "02360270",
          "02360272",
          "02360303",
          "02360323",
          "02360328",
          "02360333",
          "02360340",
          "02360346",
          "02360349",
          "02360361",
          "02360366",
          "02360371",
          "02360381",
          "02360499",
          "02360502",
          "02360503",
          "02360504",
          "02360512",
          "02360513",
          "02360524",
          "02360526",
          "02360729",
          "02360754",
          "02360757",
          "02360828",
          "02360874",
          "02360991",
          "00440167",
          "00440169",
          "01140252",
        ],
      },
      {
        "count": 0,
        "courses": [],
      },
    ],
    "science" => [
      {
        "choice": {
          "amount": 1,
          "options": Map {
            "biology_chain" => {
              "count": 1,
              "courses": [
                "01340058",
                "01340020",
              ],
            },
            "physics-chemistry_chain" => {
              "count": 1,
              "courses": [
                "01240120",
                "01140052",
              ],
            },
            "physics_chain" => {
              "choice": {
                "amount": 1,
                "options": Map {
                  "1" => {
                    "count": 1,
                    "courses": [
                      "01140075",
                    ],
                  },
                  "2" => {
                    "count": 2,
                    "courses": [
                      "01140052",
                      "01140054",
                    ],
                  },
                },
              },
            },
          },
        },
        "courses": [
          "01140075",
          "01140052",
          "01140054",
          "01140073",
          "01140101",
          "01140246",
          "01240120",
          "01250001",
          "01250801",
          "01240510",
          "01340058",
          "01340020",
        ],
        "overflow": "list_b",
        "points": 8,
      },
      {
        "choice": {
          "amount": 0,
          "options": Map {
            "biology_chain" => [
              {
                "count": 1,
                "courses": [
                  "01340058",
                  "01340020",
                ],
              },
              {
                "count": 0,
                "courses": [],
              },
            ],
            "physics-chemistry_chain" => [
              {
                "count": 1,
                "courses": [
                  "01240120",
                  "01140052",
                ],
              },
              {
                "count": 0,
                "courses": [],
              },
            ],
            "physics_chain" => [
              {
                "choice": {
                  "amount": 1,
                  "options": Map {
                    "1" => {
                      "count": 1,
                      "courses": [
                        "01140075",
                      ],
                    },
                    "2" => {
                      "count": 2,
                      "courses": [
                        "01140052",
                        "01140054",
                      ],
                    },
                  },
                },
              },
              {
                "choice": {
                  "amount": 0,
                  "options": Map {
                    "1" => [
                      {
                        "count": 1,
                        "courses": [
                          "01140075",
                        ],
                      },
                      {
                        "count": 0,
                        "courses": [],
                      },
                    ],
                    "2" => [
                      {
                        "count": 2,
                        "courses": [
                          "01140052",
                          "01140054",
                        ],
                      },
                      {
                        "count": 0,
                        "courses": [],
                      },
                    ],
                  },
                },
              },
            ],
          },
        },
        "courses": [],
        "points": 0,
      },
    ],
  },
}
`;
